Change Transfer to Submission in recordtransfer App

yenaing-oo opened the pull request:
Closes #377

- Renamed and refactored every instance of "transfer" to "submit"/"submission". This includes changes made to code, file names  and docstrings.
- In addition, I took this opportunity to make the endpoints more REST-ful (see `recordtransfer/urls.py`)
- Fixed failing E2E tests that exist in `master`

danloveg commented on app/recordtransfer/static/recordtransfer/js/submission_form/utils.js:
```diff
@@ -39,7 +39,7 @@ export const getSessionToken = () => {
 export const fetchNewSessionToken = async () => {
     try {
-        const response = await fetch("/upload-session/", {
+        const response = await fetch("/upload-session/new", {
```
This is related to my comments in urls.py, do we need the `/new` part of the URL?

yenaing-oo replied:
Addressed by: https://github.com/NationalCentreTruthReconciliation/Secure-Record-Transfer/pull/476/commits/664af7e532a4ab0bee7ccb70dce62dc703b38e54

yenaing-oo commented:
> Pretty much everything looks good to go, but I'm not so sure we should be appending `new` to the end of URLs for creating new resources. In a typical REST-ful API (which we are only partially compliant with right now), POST-ing to a endpoint with a noun name, like "upload-session/" would create an instance of that object. In this case, that would create a new upload session. I'm aware that the 'API' of this app is kind of messy right now, but I think adding `new` is going to make it _more_ messy. Glad to discuss if you feel differently.

You're right, the "new" isn't necessary. I had to refresh my understanding of REST-ful API. My reasoning was that the endpoints for making a new submission and viewing the details of a submission looked too similar, but now I see that this is OK. The other "new"s I added to make them consistent.

I've removed the "/new/" from the endpoints now.

yenaing-oo commented on app/recordtransfer/urls.py:
```diff
@@ -9,56 +9,51 @@
-        name="transfer",
+        r"^submission/new(?:/(?P<in_progress_uuid>[0-9a-f-]+))?(?:/(?P<group_uuid>[0-9a-f-]+))?/$",
+        login_required(views.pre_submission.SubmissionFormWizard.as_view()),
+        name="submit",
```
Addressed by: https://github.com/NationalCentreTruthReconciliation/Secure-Record-Transfer/pull/476/commits/777908c43f7082ce0f7c695d67a8666494708c09

yenaing-oo commented on app/recordtransfer/urls.py:
```diff
@@ -76,7 +71,7 @@
         [
             path(
-                "upload-session/",
+                "upload-session/new/",
```
Addressed by: https://github.com/NationalCentreTruthReconciliation/Secure-Record-Transfer/pull/476/commits/664af7e532a4ab0bee7ccb70dce62dc703b38e54

yenaing-oo commented:
> I'm aware that the 'API' of this app is kind of messy right now

Are there any other changes I can make to the API to address this?

danloveg commented:
> > Pretty much everything looks good to go, but I'm not so sure we should be appending `new` to the end of URLs for creating new resources. In a typical REST-ful API (which we are only partially compliant with right now), POST-ing to a endpoint with a noun name, like "upload-session/" would create an instance of that object. In this case, that would create a new upload session. I'm aware that the 'API' of this app is kind of messy right now, but I think adding `new` is going to make it _more_ messy. Glad to discuss if you feel differently.
> 
> You're right, the "new" isn't necessary. I had to refresh my understanding of REST-ful API. My reasoning was that the endpoints for making a new submission and viewing the details of a submission looked too similar, but now I see that this is OK. The other "new"s I added to make them consistent.
> 
> I've removed "/new/" from the endpoints now.

Glad to hear we're on the same page. I myself have only recently started refreshing my knowledge on REST-ful design so no worries.

danloveg commented:
> > I'm aware that the 'API' of this app is kind of messy right now
> 
> Are there any other changes I can make to the API to address this?

Most likely, yes. Though I'd like to get these changes in before picking the API apart to find inconsistencies with REST-ful design.

danloveg commented:
> 2. The views `recordtransfer:submit` and `recordtransfer:submission_detail` now overlap. If you click **Click to view report** on the profile page, you get directed to the submission form, instead of to the submission detail page. I wonder if `in_progress_uuid` and `group_uuid` should be passed as GET parameters, instead of being part of the URL?

Maybe something like this:

```
/submission?resume=uuid&group=uuid
```

yenaing-oo commented:
> 2\. The views `recordtransfer:submit` and `recordtransfer:submission_detail` now overlap. If you click **Click to view report** on the profile page, you get directed to the submission form, instead of to the submission detail page. I wonder if `in_progress_uuid` and `group_uuid` should be passed as GET parameters, instead of being part of the URL?

Now that you mention it, I've realized that the `submit` path is using a regex path to capture values in named groups. I was under the impression that this was using GET parameters, but I've realized that the reason something like this (`http://localhost:8000/submission/?in_progress_uuid=82e11932-9273-487c-909c-316aa096a31b`) works is because it's matching the uuid based on its position. Doing this ((`http://localhost:8000/submission/82e11932-9273-487c-909c-316aa096a31b`) would be equivalent.



I believe your suggestion should fix this.

> /submission?resume=uuid&group=uuid

yenaing-oo commented:
> Maybe something like this:
> 
> ```
> /submission?resume=uuid&group=uuid
> ```

Sorry this took a while. I was testing whether something like this would work:

```python
path(
        "submission/<uuid:resume>",
        login_required(views.pre_submission.SubmissionFormWizard.as_view()),
        name="resume_submit",
    ),
path(
        "submission/<uuid:group>",
        login_required(views.pre_submission.SubmissionFormWizard.as_view()),
        name="group_submit",
    ),
path(
        "submission/<uuid:uuid>/",
        login_required(views.post_submission.SubmissionDetail.as_view()),
        name="submission_detail",
    ),
path(
        "submission/",
        login_required(views.pre_submission.SubmissionFormWizard.as_view()),
        name="submit",
    ),
```

But turns out, there is no way to differentiate between the first three paths, so the first path is always chosen. The only way is to define the `submit` path as:

```python
path(
        "submission/",
        login_required(views.pre_submission.SubmissionFormWizard.as_view()),
        name="submit",
    ),
```

And pass query parameters from the templates like so:
`<a href="{% url 'recordtransfer:submit' %}?resume={{ in_progress.uuid }}">Resume</a>`

yenaing-oo commented:
> Two things I noticed:
> 
> 1. When you make a submission, the title of the page is "Transfer Complete!"
> 
> ![Screenshot 2025-03-13 131924](https://private-user-images.githubusercontent.com/28608038/422486813-f09faae0-59d7-4568-8194-82ed75743ee1.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDE4OTg0NzYsIm5iZiI6MTc0MTg5ODE3NiwicGF0aCI6Ii8yODYwODAzOC80MjI0ODY4MTMtZjA5ZmFhZTAtNTlkNy00NTY4LTgxOTQtODJlZDc1NzQzZWUxLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAzMTMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMzEzVDIwMzYxNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWM0MmI0MmM5ZTViMWEzZmNiMWY2N2M2ZDI3ODFkODllNzViNzlkNTFiZjE0MzBiNDJkYmM2MjAzOGQ5MDA1NTcmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.pXpbqjDHwqLBnhbXFcbkhr3vO8lqiGDfKebyMrBO15c)

Addressed by: https://github.com/NationalCentreTruthReconciliation/Secure-Record-Transfer/pull/476/commits/6be8c7352a5f2fc946dcc9154c57f15abad3f9fb

danloveg reviewed (approved):
Looks good, works good. Glad to have all the transfer/submission language uniform now!
